probe_src/persistent_provenance.md

```diff
@@ -0,0 +1,34 @@
+`probe_py.manual.persistent_provenance` implements persistent (between-process) provenance.
+
+`probe record ...` efficiently trackes provenance within a single process, writing the result to `probe_log`.
+If the process which reads a file is not the same as the process which writes one, but they have a common ancestor parent, this works well.
+For example, suppose a compiler reads `main.c` and writes `main.o`, and a linker reads `main.o` and writes `main.a`.
+- PROBEing just the compiler will not capture the full usage of the .c files;
+- PROBEing just the linker will not capture the full source of the .a files;
+- However PROBEing the make which invokes both (make is a common ancestor) sufficiently captures the sources and uses of all the files involved.
+
+However, there are cases where there is not a common ancestor process:
+- The computation could be could be a multi-node (a process on machine A and process on machine B have no common ancestor).
+- The computation could be carried out between restarts (process writes file, restart machine, process reads file).
+
+Therefore, we will write provenance persistently to disk in a data directory (in [XDG data
+home](https://wiki.archlinux.org/title/XDG_Base_Directory)) at transcription-time.
+
+When we PROBE a process, we learn which InodeVersions got produced, which got used, and other necessary conditions for the process to execute.
+This data will constitute the Process object.
+
+PROBE already constructs "within process-and-its-children" dataflow graph, but with the stored provenance data, we can imagine the "between processes" dataflow graph.
+The common InodeVersions between processes should get "linked up".
+
+One common query is: "what inputs got used to generate this InodeVersion?".
+The queries navigate the between-process dataflow DAG opposite the direction of information flow.
+Therefore, we need to make this operation fast.
+Scanning through every Process object will be too tedious, since ever process the user ever PROBEd will have a corresponding Process object.
+
+To speed up queries, we will have an index mapping from InodeVersions to the Processes that produced it.
+In our data directory, there will be these flat subdirectories: `inode_version_writes`, and `processes`.
+Flat directories can be thought of as key-value stores, where the filename is the key.
+- `inode_version_writes` maps an inode-and-version (filename) to the process ID (in the file contents) of the process that created it.
+- `processes` maps a process ID (filename) to a Process object (in the file contents).
+
+Note: multi-process sqlite is either not performant (globally lock entire database) or not easy (lock each row somehow), but filesystem will work fine for our case, since we just need a key-value store.
```

Comment:
sqlite has built-in serializable ACID transactions, as long as there are no network filesystems involved, we can just open the database on multiple processes and have it *just work™*; sqlite's query optimization engine is good enough we could probably just  `SELECT * FROM Processes WHERE inode_version=...`

Reply:
That being said, we probably don't want to just dump `probe_log` blobs into sqlite since it has a hardcoded limit of 1GB for strings and blobs (ask me how i know :upside_down_face:).

Reply:
Another point is that traversing the graph requires "pointer-chasing" through multiple rows (each row tells you where to find the next one). This kind of access pattern results in O(N) file reads, which may be a lot, or O(N) select-queries, which is still a lot of queries, but new query is much cheaper than open/close new file-descriptor.

Reply:
(No response required, since Ex-32 is already busy; just thinking out loud)

I need to expand on the transactions we will be doing, and at what frequency. Basically, the persistent store should be a bipartite graph between processes and inode-versions.

- At transcription-time (which probably happens no more than once per minute): insert nodes and edges.
- When the user asks for it (less than once per minute): traverse the graph "up" from a specific inode/version, "what inputs were used to make this output? (aka pull-based updating)". This query is used in applications like "Make-without-Makefile" application.
- When the user asks for it (less than once per minute): traverse the graph "down" from specific inode/version, "what outputs were dependent on this input? (aka push-based updating)". This query is used a lot less than the pull-based version, but could still be useful if the user does a system upgrade, and they don't know which of their projects need to get recompiled/recomputed.
- When the user does an SCP or Rsync (less than once per minute): traverse the graph "up", so we can transfer the "relevant" bits of provenance to the remote, so a user at the destination-machine (destination could be local or remote) can query the provenance of the files we are sending.
- When the user issues "garbage collect" (less than once per minute): iterate over every inode/version; mark the ones that still exist as "not deletable". Mark any ancestor of "not deletable" as "not deletable" (search "up" in the provenance graph from each existent provenance version).

All in all, the operations are slow enough that the parallelism between readers and writers that is potentially available in filesystem but not available in sqlite (uses readers-writers lock), is not that important. On the other hand, traversing up or down the graph is quite important. For a file-based solution, each edge-traversal requires open/close of a file; sqlite loads the graph in memory (if it fits), so each edge-traversal is a pointer dereference. If the graph does not fit in memory, sqlite will load blocks of the table (presumably LRU cached), which is still much more efficient than open/close. While a graph database (e.g., Neo4J) would be *even better*, Neo4J is not "embedable" in a Python application (only in Java). The most popular deployments of Neo4J are as a daemon process, which would be annoying. With sqlite, PROBE can be "daemonless".

Therefore, I think persistent provenance should be held in sqlite in the future. @Shofiya2003, the *logic* regarding SCP/Rsync wrapper won't change; still call `get_prov_upstream`; the output will be a set of objects, which you can call a new function, `transfer(objects, host)`, which will figure out how to transfer the objects to the remote host, whether they are filenames (the current scheme) or objects/sqlite-rows (the future scheme).

Reply:
> At transcription-time (which probably happens no more than once per minute): insert nodes and edges.

While I agree with your overall rational, some tasks, like compiling C/C++ code, where each file spawns several processes (compiler, assembler, etc.), could potentially produce higher transcription loads (in the range of once per second).

Reply:
Transcription should take place when the PROBEd process-tree completes, not when a single PROBEd process completes. I'm assuming users will do `probe make` instead of `make CC="probe gcc"` (but maybe they won't). Ideally, we need to explain that one should try to PROBE the greatest unborn ancestor process (when it gets born) :laughing: